File: app/javascript/packs/mutations/sign_in.js
import gql from "graphql-tag";

const mutation = gql`
  mutation signIn($email: String!, $password: String!) {
    signIn(email: $email, password: $password) {
      user {
        id
        email
        token
      }
    }
  }
`;

export default function signIn({ apollo, email, password }) {
  return apollo.mutate({
    mutation,
    variables: {
      email,
      password
    }
  });
}

File: app/graphql/mutations/auth/sign_in.rb
# frozen_string_literal: true

module Mutations
  module Auth
    class SignIn < Mutations::BaseMutation
      null true

      argument :email, String, required: true
      argument :password, String, required: true

      field :result, String, null: false

      def grille_resolver(email:, password:)
        user = User.find_for_authentication(email: email)
        return nil unless user

        is_valid_for_auth = user.valid_for_authentication? do
          user.valid_password?(password)
        end

        return unless is_valid_for_auth

        context[:current_user] = user
        user.as_json.slice('email', 'id').merge('token' => user.token)
      end
    end
  end
end
